peer_manager_test.go

```diff
@@ -445,51 +454,54 @@ func TestClustersPeerJoinAllAtOnce(t *testing.T) {
 	runF(t, clusters, f2)
 }
 
-func TestClustersPeerJoinAllAtOnceWithRandomBootstrap(t *testing.T) {
-	clusters, mocks := peerManagerClusters(t)
-	defer shutdownClusters(t, clusters, mocks)
-
-	if len(clusters) < 3 {
-		t.Skip("test needs at least 3 clusters")
-	}
-
-	// We have a 2 node cluster and the rest of nodes join
-	// one of the two seeds randomly
-
-	err := clusters[1].Join(clusterAddr(clusters[0]))
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	f := func(t *testing.T, c *Cluster) {
-		j := rand.Intn(2)
-		err := c.Join(clusterAddr(clusters[j]))
-		if err != nil {
-			t.Fatal(err)
-		}
-	}
-	runF(t, clusters[2:], f)
-
-	hash, _ := cid.Decode(test.TestCid1)
-	clusters[0].Pin(api.PinCid(hash))
-	delay()
-
-	f2 := func(t *testing.T, c *Cluster) {
-		peers := c.Peers()
-		if len(peers) != nClusters {
-			peersIds := []peer.ID{}
-			for _, p := range peers {
-				peersIds = append(peersIds, p.ID)
-			}
-			t.Errorf("%s sees %d peers: %s", c.id, len(peers), peersIds)
-		}
-		pins := c.Pins()
-		if len(pins) != 1 || !pins[0].Cid.Equals(hash) {
-			t.Error("all peers should have pinned the cid")
-		}
-	}
-	runF(t, clusters, f2)
-}
+// This test fails a lot when re-use port is not available (MacOS, Windows)
```

Comment:
Do we have an issue to track debugging this?  Some things to specify and/or figure out.  Did this failure effect MacOS and Windows before moving to listening on port 0, or is it the combination of using these OSs with using port 0 that causes the failures.  What line of code throws the error and what is the expected message?  Is this an error primarily coming from the way the test is written or is it reflective of a problem with ipfs-cluster?

Reply:
This is a problem with Windows and Mac not being able to use the listening port for outgoing connections, resulting in a new connection being created (with a new port) in order to bootstrap, but which cannot be used by other peers to reach that node. That, plus some operating system or Virtual Machine foo-barity. I'm happy to sacrify this test for the moment, but you're right, let's have an issue to keep in mind that it might be worth a look (and maybe enable it only in Linux): https://github.com/ipfs/ipfs-cluster/issues/306